core/src/types/list.rs

```diff
@@ -80,44 +83,81 @@ impl Stream for Lister {
             return Poll::Ready(None);
         }
 
-        if let Some(fut) = self.stating.as_mut() {
-            let (path, rp) = ready!(fut.poll_unpin(cx));
-
-            // Make sure we will not poll this future again.
-            self.stating = None;
-            let metadata = match rp {
-                Ok(rp) => rp.into_metadata(),
+        while self.task_queue.len() < self.task_queue.capacity() {
+            match ready!(self.lister.poll_next(cx)) {
+                Ok(Some(oe)) => {
+                    let (path, metadata) = oe.into_entry().into_parts();
+                    // TODO: we can optimize this by checking the provided metakey provided by services.
+                    if metadata.contains_metakey(self.required_metakey) {
+                        // NOTE: we need to maintain the order of entries, will consume the task_queue first.
+                        return if self.task_queue.is_empty() {
```

Comment:
After apply the changes of removing `buf`, I believe we don't need to perform such optimize.

Reply:
Dropped `buf` and replaced by enum `StatTask`.